packages/vite/src/node/plugins/preAlias.ts

```diff
@@ -21,7 +22,39 @@ export function preAliasPlugin(config: ResolvedConfig): Plugin {
         !options?.scan
       ) {
         if (findPatterns.find((pattern) => matches(pattern, id))) {
-          return await tryOptimizedResolve(depsOptimizer, id, importer)
+          const optimizedId = await tryOptimizedResolve(
```

Comment:
More of a question about this if statement, #8869 changed to only `tryOptimizedResolve` when an alias pattern is matched via `findPatterns`. When no patterns are matched, are we handling this elsewhere?

PS: CI is failing if you haven't noticed.

Reply:
If the alias doesn't match, then the next plugin (`@rollup/plugin-alias`) will not get into play and the normal resolution will apply (it will probably be resolved by `vite:resolve`). The problem is that when the alias match for a bare import `BI`, the `rollup/plugin-alias` will resolve it to the replacement `RE`(imagine another file in node_modules) without giving the chance to `vite:resolve` to resolve to the optimized version.

During scanning, the bare import `BI` will resolve to `RE` and be registered for optimization. The check of `tryOptimizedResolve` in preAlias will correctly resolve to the optimized version `OV` before the plugin alias can kick in before this PR.

But if we don't have the scanner anymore, then `BI` isn't yet registered as an optimized dep and the check will fail, and the alias plugin will resolve to `RE`. So what we need to do is that only for aliased bare imports, get before the alias plugin and resolve correctly to `OV` with `tryOptimizedResolve` if was already registered, or by registering as a missing import if not. It is doing what the scanner would have done, but lazily for aliases. For non-aliases, it is better to let them flow through normal resolution.

The CI is failing because VitePress is placing itself as `ssr.noExternal` but it shouldn't be optimized as it has virtual modules. We need to implement the new rules directly in this PR.

Reply:
Thanks for the explanation! I think that makes sense given we're supporting aliasing optimized deps.